pkg/virt-controller/watch/vm.go

```diff
@@ -578,6 +583,52 @@ func (c *VMController) updatePVCMemoryDumpAnnotation(vm *virtv1.VirtualMachine)
 	return nil
 }
 
+func (c *VMController) VMICPUsPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	test := fmt.Sprintf(`{ "op": "test", "path": "/spec/domain/cpu/sockets", "value": %s}`, strconv.FormatUint(uint64(vmi.Spec.Domain.CPU.Sockets), 10))
```

Comment:
why add this line ?

Reply:
> why add this line ?

@enp0s3

Reply:
@ccdlhj Hey! that was added for sake of precaution. The object could have changed while this thread processes the sync loop, and then we could end up with an error while trying to replace current value which is obsolete. Do you have objections are best practices that show the opposite? i.e. that it should not be used

Reply:
> @ccdlhj Hey! that was added for sake of precaution. The object could have changed while this thread processes the sync loop, and then we could end up with an error while trying to replace current value which is obsolete. Do you have objections or best practices that show the opposite? i.e. that it should not be used

oh.....  thank you, I have no other opinion :)

Reply:
Well, this seems redundant but in the codebase this approach is used a lot (even when not in need).
One could see it as a safeguard from forgetting to do it when it is really needed and others may consider this a waste. Either way, it is worth a discussion.

Even if someone else changes it in the meantime, you still want to override it because that is your desired value. By failing here you just delay the execution to the next reconcile iteration.

The only place where you need to test is when:
- The new value is dependent on the old value.
- The change is in a subset of the path you change (e.g. you edit one record in a list, but patch the whole list).

BTW, `add` would have worked the same as `replace` in this case. `replace` is more limiting, it requires the a previous value to exist, while `add` does not (and in most cases, you should not care about the prev).